llama_index/agent/openai_agent.py

```diff
@@ -57,267 +61,231 @@ def call_function(
+            assert isinstance(latest_function, dict)
+            self._call_function(tools, latest_function)
             n_function_calls += 1
 
-            # send function call & output back to get another response
-            all_messages = self._prefix_messages + self._memory.get()
-            chat_stream_response = StreamingAgentChatResponse(
-                achat_stream=await self._llm.astream_chat(
-                    all_messages, functions=functions
-                ),
-                sources=sources,
-            )
+        return agent_chat_response
 
-            # Get the response in a separate thread so we can yield the response
-            thread = Thread(
-                target=lambda x: asyncio.run(
-                    chat_stream_response.awrite_response_to_history(x)
-                ),
-                args=(self._memory,),
-            )
-            thread.start()
+    async def achat(
+        self,
+        message: str,
+        chat_history: Optional[list[ChatMessage]] = None,
+        *,
+        mode: ChatMode = ChatMode.default,
+    ) -> Union[AgentChatResponse, StreamingAgentChatResponse]:
+        tools, functions = self.init_chat(message, chat_history)
+        n_function_calls = 0
 
-            while chat_stream_response._is_function is None:
-                # Wait until we know if the response is a function call or not
-                time.sleep(0.05)
-                if chat_stream_response._is_function is False:
-                    return chat_stream_response
+        # Loop until no more function calls or max_function_calls is reached
+        while True:
+            agent_chat_response = await self._get_async_agent_response(mode, functions)
+            latest_function = self.session.get_latest_function_call()
+            if not self._should_continue(latest_function, n_function_calls):
+                break
+            assert isinstance(latest_function, dict)
+            self._call_function(tools, latest_function)
+            n_function_calls += 1
 
-            thread.join()
-            function_call = self._get_latest_function_call(self._memory.get_all())
+        return agent_chat_response
 
-        return chat_stream_response
+    def stream_chat(
+        self, message: str, chat_history: Optional[list[ChatMessage]] = None
+    ) -> StreamingAgentChatResponse:
+        chat_response = self.chat(message, chat_history, mode=ChatMode.stream)
+        assert isinstance(chat_response, StreamingAgentChatResponse)
+        return chat_response
+
+    async def astream_chat(
+        self, message: str, chat_history: Optional[list[ChatMessage]] = None
+    ) -> StreamingAgentChatResponse:
+        chat_response = await self.achat(message, chat_history, mode=ChatMode.stream)
+        assert isinstance(chat_response, StreamingAgentChatResponse)
+        return chat_response
 
     # ===== Query Engine Interface =====
     def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
```

Comment:
Do _query and _aqeuery still need to be defined? They look the same as the version in the base class 👀

Reply:
Good point! I didn't look much at `BaseAgent` tbh. I guess these were probably moved in a commit on main in-between when we forked and now. Took 'em out now.